T10-edit-choice-board-category

Digital-Brendan opened the pull request:
For this task, I created the functionality to create, edit, and delete categories from Firebase using new category information from the user - connecting these to the admin UI is Ali's task for this upcoming sprint.

**To do:**
You will notice there is only one test file included in this PR. As mentioned yesterday, I have been having trouble running tests, specifically with a file called `listener.dart` (if anyone has had any issues with this before or knows how to fix it, please let me know). For some reason, the one test I included doesn't cause the error but additional ones do, and in fact cause my entire codebase to crash on running.

I'm actively trying to diagnose these issues, but in the mean time, as I don't want to fall behind & Ali's task depends on this code, please wait for the additional two tests.

Commit:
Comment out broken integration test code

Commit:
Merge branch 'T05-Child-Menu-UI' into T10-Edit-Choice-Board-Category

Commit:
Comment errors

Commit:
Merge branch 'T05-Child-Menu-UI' into T10-Edit-Choice-Board-Category

Commit:
Refactor file organisation

Commit:
Merge branch 'T05-Child-Menu-UI' into T10-Edit-Choice-Board-Category

Commit:
Merge with T05

Commit:
Merge branch 'master' into T10-Edit-Choice-Board-Category
Merge with master

Commit:
Create new category feature

Commit:
Add new category to database

Commit:
Add edit category image and title functionality

Commit:
Add category delete functionality

Commit:
Test delete category functionality

Commit:
Merge with master

Commit:
Add more comments

Commit:
Remove duplicate files

Ali-Alkhars commented on lib/pages/admin/add_choice_board_category.dart:
```diff
@@ -0,0 +1,237 @@
+/// A popup card to add a new category.
+class _AddChoiceBoardCategory extends State<AddChoiceBoardCategory> {
+  File? selectedImage; // hold the currently selected image by the user
+  // controller to retrieve the user input for item name
```
edit this comment to "category" instead of item

absirgue replied:
"item name" ? do you mean category name? If not can you explain further in the comment

Ali-Alkhars commented on lib/pages/admin/add_choice_board_category.dart:
```diff
@@ -0,0 +1,237 @@
+                        icon: Icon(Icons.camera_alt),
+                        onPressed: () => pickImage(source: ImageSource.camera)),
+                    const SizedBox(height: 20),
+                    // field to enter the item name
```
update comment here too

Ali-Alkhars commented on lib/pages/admin/add_choice_board_category.dart:
```diff
@@ -0,0 +1,237 @@
+                            categoryName: categoryNameController.text),
+                        icon: Icon(Icons.add),
+                        label: const Text("Create new category"),
+                        style: const ButtonStyle(
```
Remove the button style so it doesn't override the theme colours

absirgue commented on integration_test/registration_test.dart:
```diff
@@ -1,7 +1,7 @@
 import 'package:flutter/material.dart';
 import 'package:firebase_auth/firebase_auth.dart';
-//import 'package:integration_test/integration_test.dart';
+// import 'package:integration_test/integration_test.dart';
```
Why is there this commented out import statement here ? If no reason, it's better to remove it

Digital-Brendan replied:
This isn't my code nor file but I will remove it anyway - thanks!

absirgue commented on lib/pages/admin/add_choice_board_category.dart:
```diff
@@ -0,0 +1,237 @@
+  void saveCategory(
+      {required File? image, required String? categoryName}) async {
+    if (categoryName!.isEmpty || image == null) {
+      showDialog(
```
U can create a method to show the dialog to avoid code duplication. Just create a method with the showDialog() thingy and call it with a different string each time.

-> I did that in the edit account page. Would be even better if you can reuse the method I built somehow by placing it in its all widget file or finding another nicely structured way to use the same method.

Digital-Brendan replied:
Thanks for the suggestion, though I think this is a needless complication as the amount of written code is the same in practice, if not more (function name: `showDialog()` vs `AlertDialog()`, parameters: `dialogText` vs `content`, etc). 

If you feel particularly strong about this I am absolutely happy to change it!

absirgue commented on lib/pages/admin/delete_choice_board_category.dart:
```diff
@@ -0,0 +1,57 @@
+    );
+
+    // Delete document from firebase collection
+    void deleteFromCollection(String collectionName) {
```
all of these methods (in this but also other files should be placed in a separate file that handles all method related to dealing with categories on the database

absirgue commented on lib/pages/admin/edit_choice_board_category.dart:
```diff
@@ -0,0 +1,224 @@
+  }
+
+  // Update category image and/or title
+  Future<void> editCategory(
```
this should go in the other file for example

absirgue commented on lib/pages/admin/edit_choice_board_category.dart:
```diff
@@ -0,0 +1,224 @@
+  /// Enable the user to either upload or take an image depending on
+  /// the source argument. Update the [selectedImage] with the user
+  /// provided image
+  Future pickImage({required ImageSource source}) async {
```
you should not duplicate the code, this method is in a lot of file.

absirgue commented on lib/pages/admin/edit_choice_board_category.dart:
```diff
@@ -0,0 +1,224 @@
+      ScaffoldMessenger.of(context).showSnackBar(
+        SnackBar(content: Text("No edits made")),
+      );
+    } else if (newCategoryName.isNotEmpty && newImage != null) {
```
I think this method would benefit from being refectored. It is very long. Can you try separating it in multiple functions?

Ali-Alkhars replied:
Make the try block start here and the catch becomes in place of the else. There's no need for else.

Also, show an error message in the catch block

absirgue reviewed (changes requested):
Hey!

Thanks for the work!

I added a few comments, mainly concerned with avoiding code duplication in some places and refactoring some long methods.
For the tests, you can check out Ali's branch for inspiration. He did similar testing but for items.

Thank you!

Ali-Alkhars commented on lib/widgets/add_category_button.dart:
```diff
@@ -0,0 +1,32 @@
+import 'package:seg_coursework_app/widgets/hero_dialog_route.dart';
+
+/// The button in the Admin Choice Boards page to add
+/// a new item to a category
```
update comment

Ali-Alkhars commented on lib/widgets/add_category_button.dart:
```diff
@@ -0,0 +1,32 @@
+      onPressed: addCategory,
+      icon: Icon(Icons.add),
+      label: const Text("Add a category"),
+      style: const ButtonStyle(foregroundColor: MaterialStatePropertyAll(Colors.white), backgroundColor: MaterialStatePropertyAll(Color.fromARGB(255, 80, 141, 93))),
```
Remove style

Ali-Alkhars commented on lib/pages/admin/delete_choice_board_category.dart:
```diff
@@ -0,0 +1,57 @@
+  Widget build(BuildContext context) {
+    // Return to admin screen if user cancels choice
+    Widget cancelButton = TextButton(
+      child: Text("No"),
```
Make the text here "cancel" and "delete" instead of "yes" so it becomes consistent

Ali-Alkhars commented on lib/pages/admin/delete_choice_board_category.dart:
```diff
@@ -0,0 +1,57 @@
+    // set up the AlertDialog
+    AlertDialog alert = AlertDialog(
+      title: Text("Warning!"),
+      content: Text("Are you sure you want to delete this category?"),
```
Make the text mention the category's name

Ali-Alkhars commented on lib/pages/admin/delete_choice_board_category.dart:
```diff
@@ -0,0 +1,57 @@
+        deleteFromCollection("categoryItems");
+
+        // Delete category from 'categories' collection
+        deleteFromCollection("categories");
```
The user is not sent back to admin choice boards after deleting

Digital-Brendan replied:
I'm almost sure they are - by either tapping outside the popup, tapping cancel, or tapping delete, the popup closes. It's important to remember this isn't a new screen per se, just something overlayed on screen that is then dismissed.

Ali-Alkhars replied:
I think you're missing `Navigator.pop(context);`

Ali-Alkhars commented on lib/pages/admin/add_choice_board_category.dart:
```diff
@@ -0,0 +1,237 @@
+  Widget build(BuildContext context) {
+    return Center(
+      child: Padding(
+        padding: const EdgeInsets.all(32.0),
```
change the padding value here

Digital-Brendan replied:
Change it to what?

Ali-Alkhars replied:
check the second point of the general comments I wrote. I added this line comment to mark where you should make the change to fix the text field not showing up

Ali-Alkhars reviewed (changes requested):
Hello Brendan, nice work here!

I can see that you're having issues with testing. I understand the difficulties that you're going through cause I went through them. Luckily, I found a solution. Check the test file in my branch called `firebase_functions_test.dart`. This is a unit test for my firebase functions. Basically, I used libraries which mock firebase so you could have a local firebase database and do your tests on it.

**You should also write widget tests for your widgets to test the following:**
- If all the contents e.g. buttons are there
- If clicking on buttons does the correct behaviour e.g. shows an alert if user didn't add a name or goes back to the main page if successful. 

**However,** widget tests are a bit more tricky because you need to adjust your pages to take either  real firebase functions or mock ones for testing. You could check my branch to see how I did all that.

**Also,** I will merge my branch with main very soon, so you should pull and adapt to it:
- I moved all the firebase methods into a helper class. You should add all of your firebase functions there. Lots of methods are duplicated in your branch, so this should solve the issue
- Extend the `firebase_functions_test.dart` test file to have the unit tests for your new firebase methods

### Here are also general comments (more line comments below):
- Show a snackbar message for when edits are saved
- When clicking on the category name text field the keyboard covers it. There's a simple solution to it: in the build methods of `add_choice_board_category.dart` _and edit_, change the padding into `padding: EdgeInsets.only(bottom: MediaQuery.of(context).viewInsets.bottom)` I have put a line comment in one of the files
- Deleting a category should decrement the ranks of all other categories with a rank higher than the one deleted. Check the method in my branch called `updateCategoryRanks`
- Update the text fields which ask for category names by adding these so it becomes consistent `textAlign: TextAlign.center,
                      style: TextStyle(fontSize: 25.0)`

Commit:
Uncomment lines in registration_test.dart file

Commit:
Correct comments that say item instead of category

Commit:
Remove overriding button styling

Commit:
Refactor button names, correct comments that say item, and mention category by name when deleting

Commit:
Adjust padding and text align

Commit:
Merge branch 'master' into T10-Edit-Choice-Board-Category

Commit:
Refactor to similar style as items and category items code

Commit:
Add functionality to add, edit, and delete categories

Commit:
Update category ranks on category delete

Commit:
Update tests

Commit:
Remove unneccessary functions in admin choice board

Commit:
Expand test coverage for category related functions

Commit:
Merge branch 'master' into T10-Edit-Choice-Board-Category

Commit:
Merge

Commit:
Remove unused code

Commit:
Further expand testing coverage for categories functions

Ali-Alkhars commented on lib/widgets/delete_category_button 2.dart:
```diff
@@ -0,0 +1,30 @@
+import 'package:flutter/material.dart';
```
I'm guessing you added this file by mistake?

Digital-Brendan replied:
This is a known bug with Flutter and iCloud syncing - I've removed the duplicated file

Ali-Alkhars commented on test/screens/admin/add_choice_board_category_test.dart:
```diff
@@ -0,0 +1,108 @@
+      expect(find.byType(AlertDialog), findsOneWidget);
+    });
+  });
+}
```
You should add a test which tests that adding both name and image should take user to Admin_choice_boards

Ali-Alkhars commented on test/screens/admin/delete_choice_board_category_test.dart:
```diff
@@ -0,0 +1,52 @@
+  });
+}
+
+class CollectionReferenceMock extends Mock implements CollectionReference {}
```
This class and the one below are not used. You should delete them

Ali-Alkhars commented on test/screens/admin/edit_choice_board_category_test.dart:
```diff
@@ -0,0 +1,134 @@
+
+      await tester.tap(find.byKey(ValueKey("editCategoryButton")));
+      await tester.pumpAndSettle();
+      expect(find.byType(ScaffoldMessenger), findsOneWidget);
```
Shouldn't the expected type be AlertDialog?? ScaffoldMessenger is what shows when an edit is successful

Digital-Brendan replied:
AlertDialog is correct but this is because I didn't use the `ScaffoldMessenger.of(context).showSnackBar( SnackBar(content: Text("message")), );` code that you recommended. I changed this because of the errors I was getting when executing your tests as I had pointed out the other day, though I can change it back

Ali-Alkhars replied:
I'm not sure why you get errors but having ScaffoldMessenger in the expect doesn't test what's intended

Ali-Alkhars commented on lib/helpers/firebase_functions.dart:
```diff
@@ -209,73 +161,136 @@ class FirebaseFunctions {
+
+  /// Delete category document from categories collection
+  /// Delete associated categoryItems document
+  Future deleteCategory({required String categoryId}) async {
```
Why is this a nested method?

Digital-Brendan replied:
Because it's a private function only used within the `deleteCategory()` function. While this could be written without nesting (having the code from `deleteFromCollection()` written simply within its parent function, nesting helps with readability and code reuse, all while keeping it private (it need not be accessed from elsewhere)

Ali-Alkhars commented on lib/pages/admin/delete_choice_board_category.dart:
```diff
@@ -1,54 +1,52 @@
-        // Delete category from 'categories' collection
-        deleteFromCollection("categories");
+    Widget deleteButton = TextButton(
+      child: Text("Delete"),
```
Force this button to be red, so it becomes consistent

Digital-Brendan replied:
I think I was told to remove hardcoded button styling? Though I will re-add the colouring in the mean time

Ali-Alkhars reviewed (changes requested):
There is still some work to be done. **Check these and the line comments bellow:**
- I'm guessing you didn't do manual testing! Creating a category doesn't work. The database rules have changed and now new categories need an "is_available" field. You should update the createCategory method to have new categories have `"is_available": true`
- Snackbar messages do not show to the user when editing a category. Using this instead of what you have. It should solve it `ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text("message")),
        );`
- Deleting a category doesn't works, doesn't show an error alert, doesn't pop the confirmation box
- Editing a category's name works, but editing the image does not

Digital-Brendan commented:
I'll fix the add category functionality. I'm not sure when the database rules were changed but I wasn't made aware

Digital-Brendan commented:
I'm not quite sure why some of the functionality doesn't work. Like you recommended, I took some of your firebase functions and changed them for categories

Commit:
Fix testing and bugs

Digital-Brendan commented:
Alright I implemented the changes you requested.

With regards to the error regarding editing images, it's caused by trying to delete an image that doesn't exist anymore. The code is the same as yours (almost verbatim) so could you check if you have the same issue with deleting an item. I think this is because the testing images that are hardcoded no longer exist? Other than that, I can't afford to spend any more time on this task, so I am ready to merge. @Ali-Alkhars

Ali-Alkhars commented on lib/helpers/firebase_functions.dart:
```diff
@@ -84,6 +84,7 @@ class FirebaseFunctions {
   /// Delete the image in Firestore Cloud Storage which holds
   /// the given imageUrl
   Future deleteImageFromCloud({required String imageUrl}) {
+    print(imageUrl);
```
delete debugging print

Ali-Alkhars commented on lib/helpers/firebase_functions.dart:
```diff
@@ -273,24 +274,23 @@ class FirebaseFunctions {
+    CollectionReference categories = firestore.collection('categories');
 
-    DocumentSnapshot category = await firestore.collection('categories').doc(categoryId).get();
+    DocumentSnapshot categoryItem = await categoryItems.doc(categoryId).get();
```
this line could be deleted because it's not used

Ali-Alkhars commented on test/screens/admin/edit_choice_board_category_test.dart:
```diff
@@ -107,28 +107,4 @@ void main() {
-
-      await tester.tap(find.byKey(ValueKey("editCategoryButton")));
-      await tester.pumpAndSettle();
-      expect(find.byType(ScaffoldMessenger), findsOneWidget);
```
I don't get why you removed this test? I tried inserting it back and changing `ScaffoldMessenger` to `AlertDialog` (because that's the intended behaviour) and it's successful

Digital-Brendan replied:
For some reason, the test fails for me but the reason I removed it was because, as far as I can tell, trying to edit a category that doesn't exist is impossible for the user - the categories that appear on screen are ones that exist in Firebase. The only way a user can edit a nonexistant category is by somehow specifying a categoryID, but the application doesn't provide that functionality

Digital-Brendan replied:
It's important to remember that good test coverage is about testing app functionality!

Ali-Alkhars replied:
Thank you for the lovely reminder and I do agree with you. I also believe that there could be inconsistencies with Firebase that might occur due to internet issues or any other issues, which would lead to the category being in the app but not Firebase. Therefore, a user might edit a category that is not in the database. This method tests that if such error did occur, it is caught and shown to the user.

Ali-Alkhars reviewed (changes requested):
Hello Brendan, although the changes you made did fix some of the problems, others still persist.

**Here is what I found:**
- Errors while deleting a category are not caught nor shown to the user. Nothing basically happens when an error occurs. You should catch the errors and alert the user
- Delete category only deletes the category and not the categoryItems entry. I believe that firebase doesn't allow deleting an entire collection, so you have to delete each item in the categoryItem. This should then delete the collection
- Successful category deletion doesn't show a snackbar update message to the user.
- Deleting a category doesn't update the other categories' ranks. You already wrote the function `updateAllCategoryRanks` but you didn't use it.

Commit:
Rewrite category delete function

Commit:
Implement updateAllCategoryRanks function

Commit:
Fix failing tests

Ali-Alkhars committed:
Fix category delete errors

- Add try catch block for deleting categories
- Show loading screen while deleting
- Make deleting a category also delete its image from the cloud storage
- Fix a bug in updateAllCategoryRanks
- Add missing test for edit_choice_board_category
- Fix a minor bug in edit_choice_board_category where editing only image doesn't give the category name to the method which uploads the new image

Ali-Alkhars committed:
merge master and resolve conflicts

Ali-Alkhars reviewed (approved):
Hello Brendan, so I've tested after your recent changes and there were still errors. I am really not sure why you didn't conduct these tests yourself before asking for a review.

**These were the errors/issues:**
- The `updateAllCategoryRanks` method shows an error because it's attempting to access all of the users' categories, which is not allowed by the firebase rules
- The try-catch block for deleting a category wasn't added as requested
- The edit category test which you deleted also wasn't added back
- Other minor, irrelevant issues

**Therefore,** I made the changes myself because otherwise, we would be here till the project's deadline. 

Let's not waste time and work on Sprint 3 and as agreed on our group chat, we test and make sure that we fully do our tasks before making a pull request.

Ali-Alkhars committed:
Fix errors with infinite loading screen when only editing image for categories and items

Digital-Brendan commented:
Hi Ali,

As always, thanks for the help. Just some notes on your comments:

- I'm glad you caught the error with `updateAllCategoryRanks()` though I hope you can appreciate it's extremely hard to test and catch these errors when I'm working with hardcoded category and item data. 

- Forgetting to implement the try-catch block was genuinely my fault so apologies for that. 

- As for the edit category test, as I had explained, it was a unnecessary test that inflates test coverage without testing valid functionality, though you made a great point about inconsistencies between remote and local storage - I would almost first prevent the inconsistencies and test for those before testing for editing a non-existing category.

Thanks for implementing these changes on my behalf though this isn't wasting time - I'm glad we're catching these errors before we deploy the app for our user.

Commit:
Merge branch 'master' into T10-Edit-Choice-Board-Category

Ali-Alkhars committed:
Merge master and solve conflicts

absirgue reviewed (approved):
Good job
